feat: date picker input and calendar styling

NatalieLoyzaga opened the pull request:
# Objective
This is the first piece of the DatePicker work. Which is creating a _very_ simple version of the component (bare bones API) and then the majority of the work being the styling of the input and calendar. The next piece will be getting into the meat of the API/documentation/tests side of things after that.

See entire epic for rest of work needing to be done: https://cultureamp.atlassian.net/jira/software/projects/KDS/boards/289?selectedIssue=KDS-62

# Motivation and Context
- Basic implementation of a single DatePicker with it's states styled to match the UI kit

# Notes
- Feature branch PR (with basic setup of component) is [here](https://github.com/cultureamp/kaizen-design-system/pull/2489/files)
- We've moved away from using their input and using our Kaizen TextField input. This means we'll handle the popping up of the calendar ourselves for more flexibility (which isn't in this PR). This was decided as their accessibility around the input and moving focus to the popout wasn't great and we can handle this better.
- Please disregard any API related work, this is just enough to get it rendering to style it. Although, the pattern of having it be a stateless component so we pass the date to it is how it'll work. So the story is very much how a consumer would use it.
- Label is closer in the implementation vs. the design but this matches how the TextField has it implemented. I think we'll need to update the label across the board rather than add padding to it per component. 

# Screenshots (if appropriate)
Default
![image](https://user-images.githubusercontent.com/48232362/151891725-19cf38dd-2ef3-41ba-9a41-6ae355c0ffff.png)

Default on Hover
![image](https://user-images.githubusercontent.com/48232362/151891739-cf892f9a-8930-40e2-a5df-f90f656c4576.png)

Default on Focus
![image](https://user-images.githubusercontent.com/48232362/151891766-19587078-4c35-476f-bf0a-9135ef5634de.png)

Default w/ Value
![image](https://user-images.githubusercontent.com/48232362/151891774-eba08eb7-e4ca-422f-8212-521ccbcce201.png)

Disbaled w/ Placeholder
![image](https://user-images.githubusercontent.com/48232362/151891793-9e54b316-cb62-4dc7-8f8c-9f2bea93a7cb.png)

Disabled w/ Value
![image](https://user-images.githubusercontent.com/48232362/151891802-9d923112-189b-4118-b105-12ec6d9290ac.png)

lloydstubber commented:
Updated the description to account for separated styling PRs that are now just all in here.

ActuallyACat commented:
Just a question about sequencing - which PR/issue will tackle the showing/hiding of the calender?

ActuallyACat commented on packages/date-picker/src/CalendarNav.tsx:
```diff
@@ -0,0 +1,32 @@
+import styles from "./DatePicker.scss"
+
+export type CalendarNavProps = {
+  onPreviousClick: () => void
```
/nit does `onPreviousClick` and `onNextClick` need to be documented in storybook? I can't see them in the props table.

ActuallyACat replied:
Or are these not available to the consumer? I think there's a lot of value in exposing them.

christian314159 replied:
I know naming is hard and contentious ;) but imho `onPreviousClick/onNextClick` lack some context, maybe consider something like `onPreviousMonthClick` which is bit wordier but also clearer.

lloydstubber replied:
Was leaving this kind of nitty gritty for the post-styling PR but happy to add these in as props with this PR.

lloydstubber replied:
I'm going to leave this for the next PR where we expose more of the API for consumers. This is mostly the styling/basic API to get a base in to work from.

ActuallyACat commented on packages/date-picker/src/CalendarNav.tsx:
```diff
@@ -0,0 +1,32 @@
+}) => (
+  <div className={styles.navbar}>
+    <button
+      className={styles.arrows}
```
There's nothing on these elements to indicate to a screen reader what these buttons do. Good use case for `aria-label`

lloydstubber replied:
Good catch, will update

lloydstubber commented:
@ActuallyACat that will be the next PR which will be mostly polishing of the props (e.g. disabled days) and then functionality. It was simpler to style completely first as it's easier to check with the calendar rendered beneath vs. trying to inspect when popped up.

ActuallyACat commented on packages/date-picker/src/DatePicker.tsx:
```diff
@@ -19,38 +19,34 @@ export const DatePickerWrapper: React.FunctionComponent<DatePickerProps> = ({
+        onDayClick={onDayChange}
+        navbarElement={getNavbar}
+        classNames={defaultDatePickerClasses}
+        className={styles.calendar}
```
/nit `classNames` and `className`? Could these be combined or otherwise simplified?

lloydstubber replied:
From our initial look, it's separated and we couldn't combine them but I'll take another look.

ActuallyACat commented on packages/date-picker/src/DatePicker.tsx:
```diff
@@ -19,38 +19,34 @@ export const DatePickerWrapper: React.FunctionComponent<DatePickerProps> = ({
+
+  return (
+    <>
+      <TextField
```
You'll need to be pretty careful about how people will use this component if you don't expose all these `TextField` props. Folks generally like the ability to do things like add their own automation/testing id, refs, and so on. Take a look at the [props on TextField](https://github.com/cultureamp/kaizen-design-system/blob/master/draft-packages/form/KaizenDraft/Form/TextField/TextField.tsx#L35-L50) and think about the use case for each prop and whether a consumer would like similar functionality from this component too.

christian314159 replied:
Along the same lines, should we ever have a hardcoded `id`? See `id="date-picker"` below. I can see a datepicker used multiple times on one page and then we'd have the same ID multiple times which could potentially lead to weird, unexpected behavior.

lloydstubber replied:
Thanks for that. I've passed through `id` along with the other props I think a consumer would want.

jmitsios commented:
@lloydstubber + @NatalieLoyzaga 

Looking almost perfect!   Couple small things I've noticed:
![Datepicker](https://user-images.githubusercontent.com/95595231/152709541-19fd4a24-a3c8-48e3-8cc3-dfa81ed07c10.png)
- Is there a way I can easily see what style/font/weight/size is being used in Storybook?
  - The month **title** looks a bit larger than the **Heading 4** style.
  - Are the individual dates bolder? Or is it just how it's being rendered in the browser?
- When I press Tab in the calendar to focus on elements and then select a Date with spacebar, pressing Tab once more won't jump to the next available date and/or cycle through any further.  Is this expected behaviour?
- Looks like no one designed a hover-state in the UI Kit for the Prev/Next arrows.  Could we please mirror the same hover styling used for the dates? (i.e: Blue-100 background, with Blue-500 arrow)?

christian314159 commented on packages/date-picker/src/CalendarNav.tsx:
```diff
@@ -0,0 +1,32 @@
+  onNextClick: () => void
+}
+
+export const CalendarNav: React.FunctionComponent<CalendarNavProps> = ({
```
Just wondering, is `React.FunctionComponent` the one that has a `children` prop baked in - which we wouldn't want here, would we? I think @HeartSquared always uses a different one that I can't remember.

lloydstubber replied:
This does have `PropsWithChildren` baked in it seems. Keen to know if there's one we prefer, I always reach for this.

christian314159 replied:
If we don't expect `children`, then I'd say we shouldn't have them in the type declaration. Not that it really could do much harm if someone did
```
<DatePicker><h1>Hello World</h1></DatePicker>
```
but for cleanliness' sake.

Btw, I think it's `React.VoidFunctionComponent` that could be used, but I'd also ask @HeartSquared as I know she has some views on this.

HeartSquared replied:
Yes `VoidFunctionComponent` aka `VFC` is the one without children. There was an article somewhere which mentioned that in a future major version they plan to remove children from `FC` and then deprecate `VFC`, but it hasn't been solidified yet, so in the meantime `VFC` is recommended as it is more type safe. If you do need `children`, then you would explicitly add it to your props since it adds more clarity and gives you full control of the type signature.

It's definitely on the list for Wednesday's meeting :)

lloydstubber replied:
Thanks for that, it's now been updated.

HeartSquared replied:
Here's a similar article to what I had read previously :) 
https://spin.atomicobject.com/2022/01/04/think-twice-react-fc/

christian314159 commented on packages/date-picker/src/DatePicker.tsx:
```diff
@@ -19,38 +19,34 @@ export const DatePickerWrapper: React.FunctionComponent<DatePickerProps> = ({
-        {
-          [styles.withDisabled]: isDisabled,
+}) => {
+  const getNavbar = ({ ...navbarProps }: CalendarNavProps) => (
```
tbh I'd just inline that on line 46, makes it easier see what it is without having to scan the file up/down:
```
...
getNavbar={({ ...navbarProps }: CalendarNavProps) => <CalendarNav {...navbarProps} />}
...
```

christian314159 commented on packages/date-picker/src/DatePicker.tsx:
```diff
@@ -19,38 +19,34 @@ export const DatePickerWrapper: React.FunctionComponent<DatePickerProps> = ({
+
+      <DayPicker
+        selectedDays={selectedDate}
+        // TODO create disabledDays prop, this is to display disabled styling
```
Is this TODO referring to do away with the "magic numbers" and accept a descriptive string instead? (e.g. `mon/tue/wed/...`)?

lloydstubber replied:
Correct! That's in the next card.

christian314159 reviewed (commented):
👏 Great work, just left a few general comments. Up to you to action any of them, thanks!

NatalieLoyzaga commented:
@jmitsios I have made comments for your feedback below. Let me know if there are any other issues :)

`Is there a way I can easily see what style/font/weight/size is being used in Storybook?`

Month Text: Heading-4 (1.125rem/18px font-size, 600 weight, normal spacing, 1.5rem/24px line-height) - I believe those match the UI kit so I'm unsure why it looks bigger
Day Text: Paragraph Body (1rem font-size, 400 weight, 1.5rem line-height) - The current colour seems to be just black where the UI is purple so I have updated that. 

`When I press Tab in the calendar to focus on elements and then select a Date with spacebar, pressing Tab once more won't jump to the next available date and/or cycle through any further. Is this expected behaviour?`

The expected behaviour for navigating the dates is to use the arrow keys rather than Tab. You press Tab to reach the first date then use the arrow keys to move around. 

`Looks like no one designed a hover-state in the UI Kit for the Prev/Next arrows. Could we please mirror the same hover styling used for the dates? (i.e: Blue-100 background, with Blue-500 arrow)?`

Updated now :)

jmitsios commented:
@NatalieLoyzaga — That's awesome, thanks for the helpful comments! And for updating that all so quick, it looks great.
The only thing I notice since that update is that it seems the current 'Today' date was previously Blue, now it's the same color as the other dates.  Is it okay to swap that back to Blue?  Thank you!

lloydstubber commented:
@jmitsios that style change has been updated by @NatalieLoyzaga and @ActuallyACat this is ready for re-review. I've worked through almost all the comments (and flagged some for the next stage of the work).

christian314159 commented:
Focus on previous month button:
<img width="361" alt="image" src="https://user-images.githubusercontent.com/763385/152883314-ccc52ca9-d4ab-49ab-9bd3-2c6fe93a697b.png">
I think UI Kit doesn't have the default dotted outline, just the blue one.

Focus on 6 Feb:
<img width="366" alt="image" src="https://user-images.githubusercontent.com/763385/152883123-463ae568-7030-4906-8b13-ca4adf18dba7.png">
Not sure what UI Kit suggests here, the faint outline looks too minimal to me.

@jmitsios Your thoughts?

ActuallyACat commented on packages/date-picker/src/CalendarNav.tsx:
```diff
@@ -0,0 +1,34 @@
+      className={styles.arrows}
+      onClick={() => onPreviousClick()}
+      type="button"
+      aria-label="Previous Month"
```
NB: This means that the component won't meet the `Internationalized` health metric. Not an issue at the moment as we don't have a solution for i18n in Kaizen right now. Something to keep in the back of your mind for the future.

lloydstubber replied:
Yep that's on the roadmap for the next iteration when we do the DateRange! Thanks for flagging though.

ActuallyACat reviewed (approved):
LGTM (code wise - didn't poke at the design)

lloydstubber commented:
@christian314159 What browser are you using there? I can't replicate that. This is all the states in Chrome:
<img width="367" alt="Screen Shot 2022-02-08 at 10 02 52 am" src="https://user-images.githubusercontent.com/18339250/152886881-75031f50-4f85-409e-972c-5c78190a2c21.png">

lloydstubber commented:
Ah yep that's in Firefox @christian314159. I'm guessing we'll need to prefix our focus states for that browser. Although I'm surprised we don't get this in our compile stage - @ActuallyACat is that usually the case?

<img width="382" alt="Screen Shot 2022-02-08 at 10 04 30 am" src="https://user-images.githubusercontent.com/18339250/152887055-53a4b21e-bfd7-4680-a5ac-aaf7371624eb.png">

lloydstubber commented:
@christian314159 okay that's fixed now, focus styles inside the calendar now match Chrome and FF. We just were using a mix of border vs. outline and were leveraging Chrome's inbuilt so FF was just showing it's native one.

lloydstubber commented:
## Firefox

<img width="364" alt="Screen Shot 2022-02-08 at 10 31 14 am" src="https://user-images.githubusercontent.com/18339250/152889763-f49b911d-89e2-49dc-86ac-297f7fd2eeba.png">
<img width="362" alt="Screen Shot 2022-02-08 at 10 31 08 am" src="https://user-images.githubusercontent.com/18339250/152889755-2e3d85a0-fbd1-4fa0-ac2c-2c04b4e63b4a.png">

## Chrome
<img width="359" alt="Screen Shot 2022-02-08 at 10 32 27 am" src="https://user-images.githubusercontent.com/18339250/152889828-8d302707-6e7d-47ba-a201-ede10278dc62.png">
<img width="362" alt="Screen Shot 2022-02-08 at 10 32 58 am" src="https://user-images.githubusercontent.com/18339250/152889872-b1d70770-049f-47a7-81b5-6c56e279d68c.png">

christian314159 reviewed (approved):
Looking good in Firefox now! 👍
